### DISCUSSION TITLE
Settings for git `tag` (to improve performance and readability)

### DISCUSSION THREAD

**Discussion by kuanyui** [magit/magit#4826] [Ideas]
# Issue Description
I'm using Magit on some real-worlds projects, and many projects has CD/CI, which automatically create tag for each release. But this situation critically impact the performance of Magit.

For example, on a laptop with ` Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz` + 16GB RAM
- `y y` (`magit-refs` -> `show refs, comparing them with HEAD`) to show refs: **freezing Emacs for > 5 seconds (~ 1800 tags).**
> `magit-refs` buffer, the light blue block are tags:
>
> ![Screenshot_20221213_102032](https://user-images.githubusercontent.com/1370070/207212523-44621d70-acf0-4ade-a7e8-534a72c1564a.png)

- `l l` (`magit-log` -> `current`) to list logs: I have to use `M-x toggle-truncate-lines` to see all commit message + branches, but still not practical. [See StackOverflow about this issue](https://emacs.stackexchange.com/questions/74518/how-to-hide-tag-in-magit-log?noredirect=1#comment122040_74518)

  - (`l -d l` hides all tags indeed, but it also hides all branches... That's impractical when need to do somethings on branches.)
> 
> `magit-log` buffer, the light blue block are tags.  Comparison between `M-x toggle-truncate-lines` enabled/disabled:
>
>![Screenshot_20221213_104250](https://user-images.githubusercontent.com/1370070/207214079-0e056424-2620-4611-bc1a-7e781bca2760.png)
>![Screenshot_20221213_103838](https://user-images.githubusercontent.com/1370070/207214082-04464ec5-f7b6-4788-9b1e-af35f158e7e3.png)

# Suggestions
Some ideas for possible solutions:
- `magit-refs`:
    1. Lazy-loading tags and the tags should be folded at initial state. (I guess this should be difficult according to current mechanism).
    2. A `custom` variable (which can be assigned in `.dir-locals.el`) as option to hide tags.
    3. Make separate commands / buffers for showing `branchs` only and showing `tags` only, for those who don't want to show `branches` along with `tags` in same `magit-refs` buffer.
- `magit-log`:
    1. An option / flag to totally hide tag name. (and this option must separated against `show/hide branches`, instead of `l -d l`)
    4. A `custom` variable (which can be assigned in `.dir-locals.el`) as option to show only tags count in each commit (for example, a commit has 3 tags on it, shows `(3 tags)` or so),
    5. A `custom` variable (which can be assigned in `.dir-locals.el`) as option to make a RegExp pattern to filter out (hide) some tags in logs list.

**Comment by tarsius** [comment]
I could add `--decorate-refs-exclude=<pattern>` to the log transient, but the problem with that is, that Transient and Magit do not yet support per-repository values. But you could set a local value for the `log.excludeDecoration` variable.

```
       --decorate-refs=<pattern>, --decorate-refs-exclude=<pattern>
           For each candidate reference, do not use it for decoration if it
           matches any patterns given to --decorate-refs-exclude or if it
           doesn’t match any of the patterns given to --decorate-refs. The
           log.excludeDecoration config option allows excluding refs from the
           decorations, but an explicit --decorate-refs pattern will override
           a match in log.excludeDecoration.
```

**Reply by kuanyui** [reply]
1. `magit-log`: Sounds good. I think a local variable which can be used in `.dir-locals.el` under the root directory of git repo is enough.
2. `magit-refs`: Does the `--decorate-refs`also workable for `magit-refs`? In my case, `magit-refs` spends more time than `magit-log`...

**Reply by kyleam** [reply]
> 2\. Does the `--decorate-refs`also workable for `magit-refs`?

No, but you could do something like this to disable tags until you want to see them:

```elisp
(defun my/magit-refs-toggle-tags ()
  "Toggle showing tags in `magit-refs-mode'.
This only affects the current buffer and is useful if you do not
show tags by default."
  (interactive)
  (let ((pos (point)))
    (if (memq 'magit-insert-tags magit-refs-sections-hook)
        (kill-local-variable 'magit-refs-sections-hook)
      (setq-local magit-refs-sections-hook
                  (append magit-refs-sections-hook
                          '(magit-insert-tags))))
    (magit-refresh-buffer)
    (goto-char pos)))

(with-eval-after-load 'magit-refs
  (remove-hook 'magit-refs-sections-hook #'magit-insert-tags)
  (define-key magit-refs-mode-map (kbd "C-c C-t") #'my/magit-refs-toggle-tags))
```

**Reply by kuanyui** [reply]
@kyleam Thanks for this hack! Works great and blazing fast! (Sorry for late replying)

**Reply by kuanyui** [reply]
@tarsius If there's any news on it? Thanks
